crates/iota-indexer/src/handlers/mod.rs

```diff
@@ -45,3 +53,116 @@ pub struct EpochToCommit {
     pub new_epoch: IndexedEpochInfo,
     pub network_total_transactions: u64,
 }
+
+pub struct CommonHandler<T> {
+    handler: Box<dyn Handler<T>>,
+}
+
+impl<T> CommonHandler<T> {
+    pub fn new(handler: Box<dyn Handler<T>>) -> Self {
+        Self { handler }
+    }
+
+    async fn start_transform_and_load(
+        &self,
+        cp_receiver: iota_metrics::metered_channel::Receiver<(u64, T)>,
+        cancel: CancellationToken,
+    ) -> IndexerResult<()> {
+        let checkpoint_commit_batch_size = std::env::var("CHECKPOINT_COMMIT_BATCH_SIZE")
+            .unwrap_or(CHECKPOINT_COMMIT_BATCH_SIZE.to_string())
+            .parse::<usize>()
+            .unwrap();
+        let mut stream = iota_metrics::metered_channel::ReceiverStream::new(cp_receiver)
+            .ready_chunks(checkpoint_commit_batch_size);
+
+        let mut unprocessed = BTreeMap::new();
+        let mut tuple_batch = vec![];
+        let mut next_cp_to_process = self
+            .handler
+            .get_watermark_hi()
+            .await?
+            .map(|n| n.saturating_add(1))
+            .unwrap_or_default();
+
+        loop {
+            if cancel.is_cancelled() {
+                return Ok(());
+            }
+
+            // Try to fetch new data tuple from the stream
+            if unprocessed.len() >= UNPROCESSED_CHECKPOINT_SIZE_LIMIT {
+                tracing::info!(
+                    "Unprocessed checkpoint size reached limit {}, skip reading from stream...",
+                    UNPROCESSED_CHECKPOINT_SIZE_LIMIT
+                );
+            } else {
+                // Try to fetch new data tuple from the stream
+                match stream.next().now_or_never() {
+                    Some(Some(tuple_chunk)) => {
+                        if cancel.is_cancelled() {
+                            return Ok(());
+                        }
+                        for tuple in tuple_chunk {
+                            unprocessed.insert(tuple.0, tuple);
+                        }
+                    }
+                    Some(None) => break, // Stream has ended
+                    None => {}           // No new data tuple available right now
+                }
+            }
+
+            // Process unprocessed checkpoints, even no new checkpoints from stream
+            let checkpoint_lag_limiter = self.handler.get_max_committable_checkpoint().await?;
```

Comment:
if this fails, then in this task will fail silently without any errors in the `start_objects_snapshot_handler` function. Is this expected?

```rust
spawn_monitored_task!(common_handler.start_transform_and_load(receiver, cancel));
```

Reply:
Created issue to investigate this https://github.com/iotaledger/iota/issues/5780